src/Http/Routing/src/Builder/EndpointRoutingApplicationBuilderExtensions.cs

```diff
@@ -96,10 +106,22 @@ public static IApplicationBuilder UseEndpoints(this IApplicationBuilder builder,
             //
             // Each middleware gets its own collection of data sources, and all of those data sources also
             // get added to a global collection.
-            var routeOptions = builder.ApplicationServices.GetRequiredService<IOptions<RouteOptions>>();
-            foreach (var dataSource in endpointRouteBuilder.DataSources)
+            // In the global endpoint route case we only want to copy data sources once, so we wait for a specific key before copying data sources
+            // like in the case of minimal hosting
+            if (builder.Properties.TryGetValue(GlobalEndpointBuilderCopyRoutesKey, out _) ||
```

Comment:
This feels so weird to me. I wonder if we should just not add duplicates to the global collection....